Greatest Common  Divisor function

duhan9836 opened the pull request.

duhan9836 committed:
Greatest Common  Divisor function

duhan9836 committed:
Update GCD.py

add euclid's algorithm to solve the same problem.

duhan9836 committed:
Update GCD.py

amend a minor error

duhan9836 committed:
FractionClass

duhan9836 committed:
gatecircuitsclass

caitao1985 commented on Fraction_class.py:
```diff
@@ -0,0 +1,87 @@
+def gcd_euclid_recursion(m,n):
+    m=abs(m)
+    n=abs(n)
+    if m==n:
```
logically this is not necessary, as if `m == n`, m % n == 0, covered by logic below.

duhan9836 replied:
make sense!

caitao1985 commented on Fraction_class.py:
```diff
@@ -0,0 +1,87 @@
+        temp=m%n
+        m=n
+        n=temp
+        euc=gcd_euclid_recursion(m,n)
```
should be written as `return gcd_euclid_recursion(m,n)`, tail recursive. 👍

duhan9836 replied:
what does tail recursive mean?

caitao1985 commented on GCD.py:
```diff
@@ -0,0 +1,53 @@
+    else:
+        i=n-1
+        while i>1:
+            if m%i==0 and n%i==0:
```
alternatively, this can be re-written as:

```python
       for i in range(n-1, 1, -1)：
            if m%i==0 and n%i==0:
                return i
        return 1
```

caitao1985 commented on GCD.py:
```diff
@@ -0,0 +1,53 @@
+        return n
+    else:
+        while m%n!=0:
+            temp=m%n
```
```python
            temp=m%n
            m=n
            n=temp
```
equal to:

```python
            m, n=n, m%n
```

duhan9836 replied:
tuple unpack. zan!

duhan9836 committed:
alternative ways of finding the kth item in a list

first function simply sort the list and find the kth.
second function implement half way through the quick sort and find the kth by comparing the pivot index with the target index.

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+                j=j-1
+                if a[j]<pivot:
+                    a[i]=a[j]
+                    a[j]=pivot
```
`a[j]=pivot`, this line can be removed.

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+                    a[j]=pivot
+                    break
+        i=i+1
+    if a[i-1]>pivot:
```
```python
    if a[i-1]>pivot:
        a[i-1]=pivot
        return i-1
    else:
        return j
```
should be replaced with

```python
return j
```

duhan9836 commented:
多个else有啥问题吗？

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: tao <notifications@github.com>
Sent: Tuesday, December 10, 2019 6:25:00 PM
To: duhan9836/Algorithm-learning <Algorithm-learning@noreply.github.com>
Cc: duhan9836 <han.du.o@outlook.com>; Author <author@noreply.github.com>
Subject: Re: [duhan9836/Algorithm-learning] Greatest Common Divisor function (#10)


@caitao1985 commented on this pull request.

________________________________

In findkth.py<https://github.com/duhan9836/Algorithm-learning/pull/10#discussion_r355955230>:

> +#so for each sub list, if the right place for pivot is just the kth, we know the answer would be the pivot then.

+#so we 分解 the question into smaller question, as to for each pivot, find its right place, and compare the place with k.

+

+def findp(a,i,j):

+    pivot=a[j]

+    while i<j:

+        if a[i]>pivot:

+            a[j]=a[i]

+            while j>i+1:

+                j=j-1

+                if a[j]<pivot:

+                    a[i]=a[j]

+                    a[j]=pivot

+                    break

+        i=i+1

+    if a[i-1]>pivot:



    if a[i-1]>pivot:

        a[i-1]=pivot

        return i-1

    else:

        return j

should be replaced with

return j

―
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/duhan9836/Algorithm-learning/pull/10?email_source=notifications&email_token=AN2CUWV7O4TIPUJQLMN7UFLQX5U7ZA5CNFSM4JVG6ELKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOTUH6I#pullrequestreview-329729017>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AN2CUWQPRRA6NTUAHCRYAHLQX5U7ZANCNFSM4JVG6ELA>.

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+
+def findp(a,i,j):
+    pivot=a[j]
+    while i<j:
```
my alternative implementation 😄 
```python
    while i<j:
        if a[i]<=pivot:
            i +=1
            continue
        a[j]=a[i]
        j-=1
        while i < j and a[j] >= pivot: j-=1
        if i < j:
            a[i] = a[j]
            i += 1
```

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+#print(findp(b,0,len(b)-1))
+#print(b)
+
+def find(a,k,i,j):
```
👍 !!!

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+    p=findp(a,i,j)
+    if p==k-1:
+        return a[p]
+    else:
```
```python
    else:
        if p>k-1:
            j=p-1
        else:
            i=p+1
    return find(a,k,i,j)
```
this can be replaced with
```python
seg = (i, p-1) if p > k-1 else (p+1, j)
return find(a,k,*seg)
```

`seg` expanded to i, j when calling find.

caitao1985 commented on findkth.py:
```diff
@@ -0,0 +1,61 @@
+        return a[p]
+    else:
+        if p>k-1:
+            j=p-1
```
k need to be adjusted as well. 😃

duhan9836 commented:
seg是啥？

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: tao <notifications@github.com>
Sent: Tuesday, December 10, 2019 6:43:32 PM
To: duhan9836/Algorithm-learning <Algorithm-learning@noreply.github.com>
Cc: duhan9836 <han.du.o@outlook.com>; Author <author@noreply.github.com>
Subject: Re: [duhan9836/Algorithm-learning] Greatest Common Divisor function (#10)


@caitao1985 commented on this pull request.

________________________________

In findkth.py<https://github.com/duhan9836/Algorithm-learning/pull/10#discussion_r355964157>:

> +        a[i-1]=pivot

+        return i-1

+    else:

+        return j

+b=[26, 42, 38, 33, 48, 58, 77, 76, 78, 98]

+

+#print(findp(b,0,len(b)-1))

+#print(b)

+

+def find(a,k,i,j):

+    p=findp(a,i,j)

+    if p==k-1:

+        return a[p]

+    else:

+        if p>k-1:

+            j=p-1



k need to be adjusted as well. 😃

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/duhan9836/Algorithm-learning/pull/10?email_source=notifications&email_token=AN2CUWTREQZNFDJXH7DMAQLQX5XFJA5CNFSM4JVG6ELKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOTXBJQ#pullrequestreview-329740454>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AN2CUWQLNLYZR4BEVQ64IKLQX5XFJANCNFSM4JVG6ELA>.

duhan9836 commented:
我特意adjust了卡，因为题目是说第k个，就是indexk-1

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Han Du <han.du.o@outlook.com>
Sent: Tuesday, December 10, 2019 6:44:14 PM
To: duhan9836/Algorithm-learning <reply@reply.github.com>; duhan9836/Algorithm-learning <Algorithm-learning@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: Re: [duhan9836/Algorithm-learning] Greatest Common Divisor function (#10)

seg是啥？

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: tao <notifications@github.com>
Sent: Tuesday, December 10, 2019 6:43:32 PM
To: duhan9836/Algorithm-learning <Algorithm-learning@noreply.github.com>
Cc: duhan9836 <han.du.o@outlook.com>; Author <author@noreply.github.com>
Subject: Re: [duhan9836/Algorithm-learning] Greatest Common Divisor function (#10)


@caitao1985 commented on this pull request.

________________________________

In findkth.py<https://github.com/duhan9836/Algorithm-learning/pull/10#discussion_r355964157>:

> +        a[i-1]=pivot

+        return i-1

+    else:

+        return j

+b=[26, 42, 38, 33, 48, 58, 77, 76, 78, 98]

+

+#print(findp(b,0,len(b)-1))

+#print(b)

+

+def find(a,k,i,j):

+    p=findp(a,i,j)

+    if p==k-1:

+        return a[p]

+    else:

+        if p>k-1:

+            j=p-1



k need to be adjusted as well. 😃

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/duhan9836/Algorithm-learning/pull/10?email_source=notifications&email_token=AN2CUWTREQZNFDJXH7DMAQLQX5XFJA5CNFSM4JVG6ELKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOTXBJQ#pullrequestreview-329740454>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AN2CUWQLNLYZR4BEVQ64IKLQX5XFJANCNFSM4JVG6ELA>.
